streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

```diff
@@ -666,16 +672,16 @@ public void shouldCloseTaskAsZombieAndRemoveFromActiveTasksIfProducerWasFencedWh
             new TestCondition() {
                 @Override
                 public boolean conditionMet() {
-                    return producer.commitCount() == 2;
+                    return producer.commitCount() == 1;
```

Comment:
Why did this go from 2 to 1? other than not passing an arg to `runOnce` the test logic to this point hasn't changed

Reply:
The logic does have changed: in the old code we will commit twice on producer, one during the rebalance and one from the elapsed time. In the new code, the optimization I added will realize that nothing has been generated since the last commit, and hence we will skip committing in this case. 

Thinking about it, this does have a side-effect though since for EOS if commit was not called in a long time then txn will be aborted, and if producer does not talk to txn coordinator even longer it could be removed as well. But personally I think it is okay for such scenario to happen, since really no data was generated, and hence committing an empty txn does not really make sense, and we should rather increase the txn expiration time in this case. WDYT? @mjsax @vvcephei

Reply:
I guess that we always have a transaction open, not just when we have something to commit.

It seems like one solution is to open a transaction only when we have data to process. Although this might complicate things.

Alternatively, is there a way to periodically send a "keep alive" message to let the broker know we do still intend to use that transaction? It seems like either this or just abort/close the empty txn and re-open is better than a super-long expiration time. Otherwise, why is there even an expiration time?

Is there any tradeoff between having one transaction open for a super long time, vs periodically closing empty transactions and starting new ones?

Reply:
Completing a txn and starting a new one come with some cost, and hence is what we want to avoid generally.

On the other hand, we do not yet have a mechanism for "keep alive": with that, I think keeping a long lived EMPTY txn is okay, note that if the txn is not empty, then not committing it in time will increase the latency. Hence I'm only trying to optimize the case when the txn is empty.

Reply:
I just talked to @hachikuji about this. Not committing is actually fine. Note, that beginTx() is a client local state transition -- nothing is written to the log (there are not "begin tx markers") and the TC state is also not modified. This implies, that the transaction timeout is not started on beginTx() -- the timeout only starts after the first record was written to the log. Thus, we don't even need to tell users to increase the tx timeout for low traffic topic that might have longer periods with no data.